### ISSUE TITLE
Segmentation fault at the end of training 

### ISSUE THREAD

**Issue by iamyangchen** [horovod/horovod#2208]
**Environment:**
1. Framework: (TensorFlow, Keras, PyTorch, MXNet)
Tensorflow

2. Framework version:
1.15-gpu

3. Horovod version:
Custom built: v0.19.2 plus cherry-picked https://github.com/horovod/horovod/pull/1830
(Seems latest release still doesn't contain the tf join op, so I cherry-picked the change to v0.19.2, the version I have used before)

4. MPI version:
4.0.0

5. CUDA version:
6. NCCL version:
CUDNN_VERSION=7.6.5.32-1+cuda10.1
NCCL_VERSION=2.4.8-1+cuda10.1

7. Python version:
2.7

8. OS and version:
Base docker is tensorflow/tensorflow:1.15.0-gpu

9. GCC version:
4.8

**Your question:**
Please ask your question here.

I trained on 8 workers, each worker has 1 gpu. When creating the MonitoredTrainingSession I added a session hook that runs the hvd.join() op at the end of the session according to the suggestion from https://github.com/horovod/horovod/pull/1830. At the end of the training, occasionally I got this segmentation fault:

```
*** Process received signal ***
Signal: Segmentation fault (11)
Signal code: Address not mapped (1)
Failing at address: 0x55b000000000
[ 0] /lib/x86_64-linux-gnu/libc.so.6(+0x3ef20)[0x7f3769d60f20]
[ 1] /usr/lib/x86_64-linux-gnu/libtcmalloc.so.4(_ZN8tcmalloc11ThreadCache21ReleaseToCentralCacheEPNS0_8FreeListEmi+0xdb)[0x7f376a137bcb]
[ 2] /usr/lib/x86_64-linux-gnu/libtcmalloc.so.4(_ZN8tcmalloc11ThreadCache11ListTooLongEPNS0_8FreeListEm+0x1b)[0x7f376a137c9b]
[ 3] /usr/lib/x86_64-linux-gnu/libtcmalloc.so.4(cfree+0x2d5)[0x7f376a1466f5]
[ 4] /usr/local/lib/python2.7/dist-packages/horovod/tensorflow/mpi_lib.so(_ZNSt10_HashtableISsSt4pairIKSsSt6vectorIN7horovod6common7RequestESaIS5_EEESaIS8_ENSt8__detail10_Select1stESt8equal_toISsESt4hashISsENSA_18_Mod_range_hashingENSA_20_Default_ranged_hashENSA_20_Prime_rehash_policyENSA_17_Hashtable_traitsILb1ELb0ELb1EEEE5eraseENSA_20_Node_const_iteratorIS8_Lb0ELb1EEE+0x10a)[0x7f362a66ac2a]
[ 5] /usr/local/lib/python2.7/dist-packages/horovod/tensorflow/mpi_lib.so(_ZN7horovod6common10Controller17ConstructResponseERSsi+0xb50)[0x7f362a65f740]
[ 6] /usr/local/lib/python2.7/dist-packages/horovod/tensorflow/mpi_lib.so(_ZN7horovod6common10Controller19ComputeResponseListERSt6atomicIbERNS0_18HorovodGlobalStateE+0x1825)[0x7f362a6652a5]
[ 7] /usr/local/lib/python2.7/dist-packages/horovod/tensorflow/mpi_lib.so(+0x6094b)[0x7f362a68194b]
[ 8] /usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0xbd66f)[0x7f3768c5a66f]
[ 9] /lib/x86_64-linux-gnu/libpthread.so.0(+0x76db)[0x7f3769b0a6db]
10] /lib/x86_64-linux-gnu/libc.so.6(clone+0x3f)[0x7f3769e4388f]
```

The training takes hours, and this issue doesn't always happen (e.g., my shuffling of training dataset might be part of the reason). I'm wondering if experts here can help point out a direction in troubleshooting the issue.

**Comment by tgaddair** [comment]
@Richie-yan have you observed this before?

**Comment by Richie-yan** [comment]
@iamyangchen Sorry, I have not encountered this situation before. May I ask, what is the probability of this error? Can it be reproduced in a specific situation?

**Comment by Line290** [comment]
I had the same problem. I trained a network on 1 worker with 8 GPUs. After 62k mini-batch, it stopped and throw out "rank0 segmentation fault". Then, I retrain the model, the error seems to disappear.

**Comment by Richie-yan** [comment]
Hi, @Line290 After 62k mini-batch, are you sure that a certain rank has finished reading the data at this time? I want to first determine whether to start calling join logic at this time?

**Comment by Line290** [comment]
Yes, each rank has ~112k examples, mini-batch is 4 for a single GPU.

**Comment by Line290** [comment]
Yes
